xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/macros.vm

```diff
@@ -65,6 +65,21 @@
 #macro(submitButton $name $shortcut $value $class)
 <input class="btn $!class" type="submit" name="$name"#if($keyboardShortcutsEnabled) title="$shortcut"#end value="$value"/>
 #end
+#**
+ * Returns the escaped value of the shortcut. This value can be set in two ways: via user preferences and wiki level translation
+ * keys. If there are both, the user preference will take priority.
+ * @param userPreferenceName The id of the user preference that is listed in XWikiUsersDocumentInitializer.java
+ *        e.g. `shortcut_developer_usertype`
+ * @param translationKey  The key of the translation containing the wiki level defined shortcut.
+ *        e.g. `core.shortcuts.developer.user.type`
+ * @since 16.9.0-rc-1
+ *#
+#macro(getShortcutValue $translationKey)
+#set ($shortcutPreferenceObject = $xwiki.getDocument($xcontext.getUser()).getObject('XWiki.UserPreferenceShortcutsClass', 'translationKey', $translationKey))
```

Comment:
This call is quite expensive. This gets the document from the cache and performs a full clone of it. This seems particularly dangerous as this macro is likely called in a loop with all shortcuts and this might happen in every request (not sure if there is further caching).

I wonder if it wouldn't make sense to introduce a script service for it. This could help in two ways:

1. It could avoid the clone of the document.
2. It/a separate component could contain a cache of the shortcuts, making it even cheaper to get all shortcuts.

An alternative to an extra cache could also be a way to simply get all shortcuts in the script service to avoid repeatedly calling the method when a larger number of shortcut values is needed.

Reply:
> I wonder if it wouldn't make sense to introduce a script service for it. This could help in two ways:
>    * It could avoid the clone of the document.
>    * It/a separate component could contain a cache of the shortcuts, making it even cheaper to get all shortcuts.


Isn't the main reason for it to be cheaper that we avoid cloning the doc? AFAIK 1. is not an issue except for the performance gain described in 2. 

> An alternative to an extra cache could also be a way to simply get all shortcuts in the script service to avoid repeatedly calling the method when a larger number of shortcut values is needed.

:+1: If we set up a script service, might as well make the most of it and make ourselves some shortcuts :)
From what I understood, it could also be done in addition to a cache.

Reply:
Talking about performance without having actual numbers (from profiling) is hard. It's a feeling of me that this could impact the performance, but it would be best to perform some actual profiling to see how bad it really is. I assume that adding a script service to get all shortcuts should already be a huge win. Adding a cache then adds quite some complexity that might not be warranted. A possibility could also be to cache shortcuts just for the current request in the context.

Reply:
When looking on making a script service, I found `$crtUserDoc`. It's in our velocity scripting doc and from what I understand, it would solve most of the performance issue with this. Currently, it's not used at all in XS, but it seems like our situation here is exactly why it's available.

https://github.com/xwiki/xwiki-platform/blob/bc80e8eac2dedfc11a4b9201b6517c50756dce1f/xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/main/resources/templates/frequentlyUsedDocs.vm#L29

Testing it quickly.

This'd mean that there's still one `getObject` call for each shortcut, but as far as I understand, this is not as much of a performance issue as the `getDocument`.

Reply:
Proposed the improvement in https://github.com/xwiki/xwiki-platform/pull/2981/commits/bbc63a04652dd86aeec049ed67bb2336ade88bb1 .
Do you think a script service is still useful in light of this new knowledge?

Reply:
While a script service could certainly make things a bit cleaner, the performance should indeed be okay that way. The clone happens only once this way the first time an object is accessed.